src/operations/execute_operation.ts

```diff
@@ -170,60 +160,60 @@ function executeWithServerSelection(
-      return callback(undefined, result);
-    }
-
-    const hasReadAspect = operation.hasAspect(Aspect.READ_OPERATION);
-    const hasWriteAspect = operation.hasAspect(Aspect.WRITE_OPERATION);
-    const itShouldRetryWrite = shouldRetryWrite(err);
-
-    if ((hasReadAspect && !isRetryableError(err)) || (hasWriteAspect && !itShouldRetryWrite)) {
-      return callback(err);
-    }
+  function retryOperation(originalError: MongoError, maxWireVersion: number) {
+    const isWriteOperation = operation.hasAspect(Aspect.WRITE_OPERATION);
+    const isReadOperation = operation.hasAspect(Aspect.READ_OPERATION);
 
-    if (
-      hasWriteAspect &&
-      itShouldRetryWrite &&
-      err.code === MMAPv1_RETRY_WRITES_ERROR_CODE &&
-      err.errmsg.match(/Transaction numbers/)
-    ) {
-      callback(
+    if (isWriteOperation && originalError.code === MMAPv1_RETRY_WRITES_ERROR_CODE) {
+      return callback(
         new MongoServerError({
           message: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
           errmsg: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
-          originalError: err
+          originalError
         })
       );
+    }
 
-      return;
+    if (isWriteOperation && !isRetryableWriteError(originalError, maxWireVersion)) {
+      return callback(originalError);
     }
 
-    // select a new server, and attempt to retry the operation
-    topology.selectServer(selector, serverSelectionOptions, (e?: any, server?: any) => {
-      if (
-        e ||
-        (operation.hasAspect(Aspect.READ_OPERATION) && !supportsRetryableReads(server)) ||
-        (operation.hasAspect(Aspect.WRITE_OPERATION) && !supportsRetryableWrites(server))
-      ) {
-        callback(e);
-        return;
-      }
+    if (isReadOperation && !isRetryableReadError(originalError)) {
+      return callback(originalError);
+    }
 
+    if (
+      originalError instanceof MongoNetworkError &&
+      session.isPinned &&
+      !session.inTransaction() &&
+      operation.hasAspect(Aspect.CURSOR_CREATING)
+    ) {
       // If we have a cursor and the initial command fails with a network error,
       // we can retry it on another connection. So we need to check it back in, clear the
       // pool for the service id, and retry again.
-      if (
-        err &&
-        err instanceof MongoNetworkError &&
-        server.loadBalanced &&
-        session &&
-        session.isPinned &&
-        !session.inTransaction() &&
-        operation.hasAspect(Aspect.CURSOR_CREATING)
-      ) {
-        session.unpin({ force: true, forceClear: true });
+      session.unpin({ force: true, forceClear: true });
+    }
+
+    // select a new server, and attempt to retry the operation
+    topology.selectServer(selector, serverSelectionOptions, (error?: Error, server?: Server) => {
+      if (!error && operation.hasAspect(Aspect.READ_OPERATION)) {
+        return callback(
+          new MongoUnexpectedServerResponseError('Selected server does not support retryable reads')
```

Comment:
This block should be deleted: I think you removed the `!supportsRetryableReads` check which would always return false, so L199 would have been equivalent to an `... && false`, i.e., never happens; as is, the current logic will always throw on retrying reads

Reply:
I suppose we should keep that feature

Reply:
😂